docs/proposals/kubelb.md

```diff
@@ -0,0 +1,197 @@
+# KubeLB
+
+**Author**: Matthias Osthues
+
+**Status**: Draft proposal; prototype in progress.
+
+## Goals
+
+KubeLB is a provider for multi cluster load balancing and takes the advantages of Kubernetes to do the load balancing. 
+
+## Non-Goals
+
+* Implementation of service type LoadBalancer for single cluster
+
+* ECMP and BGP
+
+* IPv6 support
+
+## Motivation and Background
+
+Kubernetes does not offer an out of the box implementation of load balancers for clusters. The implementations of Network LB & Ingress LB that Kubernetes does ship with are all calls out to various IaaS platforms (GCP, AWS, Azure…). If you’re not running on a supported IaaS platform (GCP, AWS, Azure…), LoadBalancers & Ingress will not get provisioned.
+Solutions which are available e.g. MetalLB focus on a single cluster. KubeLB aims to provide load balancing for multiple clusters and takes the advantages of kubernetes itself. 
+ 
+#### Possible features: 
+
+* Implementation of service type LoadBalancer where it's not available
+
+* Multi cluster Ingress Controller
+
+* Multi cluster load balancing
+
+* Possible cost reduction as you can decide between KubeLB and IaaS LoadBalancer instances
+
+* Load balancing for internal networks
+ 
+* In a multi cluster environment it can be useful to have a limited amount of entrypoints (Load balancer cluster) inside the network to do some security or monitoring. 
+
+* Make the actual loadbalancer technology pluggable (easy implementation if e.g. someone wants to use nginx over envoy)
+
+* Be able to have the load balancing cluster exist separately from the seed cluster or integrated into it
+
+## Implementation
+
+The overall implementation contains two different parts: 
+
+**Agent**: Controller which is deployed in every tenant cluster. It watches for Services, Ingress resources and node changes. Updates or Creates a CRD inside the load balancing Cluster accordingly.
+
+**Manager**: Controller which is responsible for deploying and configuring the resources needed inside the load balancer cluster. Watcher for it's CRD which describes the load balancing endpoints and settings. 
+
+<div style="text-align:center">
+  <img src="images/kubelb-architecture.png" width="100%" />
+</div>
+
+
+It is required to have a service type "LoadBalancer" implementation on the load balancer cluster. This can be a cloud solution, or some single cluster implementations. 
+
+#### Implementation L4
+
+The agent watches for Services of type LoadBalancer/NodePort in the tenant cluster. In the tenant cluster itself for each service with type LoadBalancer a NodePort is allocated by default. 
+The agent informs the Manager which creates a Service and Endpoint in the LB cluster and adds the node IP addresses of the tenant cluster to the Endpoint IP addresses there. The agent watches for node changes like "remove", "add" and failures and will update the IP list in the Endpoint accordingly.
```

Comment:
The usage of node IP addresses here makes an assumption on having them reachable from the LB cluster. This works if the LB cluster and the tenant clusters are part of the same admin domain (infrastructure), or if the tenant cluster nodes have public IPs.

Potentially, we could introduce some tunneling mechanism between the LB and the tenant clusters, which would also allow for:
- running tenant clusters behind NAT / firewall (e.g. LB cluster in a public cloud and tenant clusters on premises or on different cloud providers)
- encryption of the traffic between the LB and the tenant clusters (may be helpful especially if TLS sessions are terminated in the LB cluster and / or if public internet is used to interconnect LB and tenant clusters)

Reply:
Thanks for pointing out this as a feature. If i have some time left i will have a look at this. I am a big fan of Wireguard which could be a potential easy implementation for this feature.